src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h

```diff
@@ -86,40 +87,64 @@ class GcsPlacementGroupSchedulerInterface {
 class ScheduleContext {
  public:
   ScheduleContext(std::shared_ptr<absl::flat_hash_map<NodeID, int64_t>> node_to_bundles,
-                  const absl::optional<std::shared_ptr<BundleLocations>> bundle_locations,
-                  const GcsNodeManager &node_manager)
+                  const absl::optional<std::shared_ptr<BundleLocations>> bundle_locations)
       : node_to_bundles_(std::move(node_to_bundles)),
-        bundle_locations_(bundle_locations),
-        node_manager_(node_manager) {}
+        bundle_locations_(bundle_locations) {}
 
   // Key is node id, value is the number of bundles on the node.
   const std::shared_ptr<absl::flat_hash_map<NodeID, int64_t>> node_to_bundles_;
   // The locations of existing bundles for this placement group.
   const absl::optional<std::shared_ptr<BundleLocations>> bundle_locations_;
-
-  const GcsNodeManager &node_manager_;
 };
 
 class GcsScheduleStrategy {
  public:
+  GcsScheduleStrategy(GcsResourceManager &gcs_resource_manager)
+      : gcs_resource_manager_(gcs_resource_manager) {}
   virtual ~GcsScheduleStrategy() {}
   virtual ScheduleMap Schedule(
       std::vector<std::shared_ptr<ray::BundleSpecification>> &bundles,
       const std::unique_ptr<ScheduleContext> &context) = 0;
+
+ protected:
+  /// Reset acquired resources.
+  ///
+  /// \return Void.
+  void ResetAcquiredResources();
+
+  /// Records the acquirement of resources from the specified node.
+  ///
+  /// \param node_id Id of a node.
+  /// \param required_resources Resources to apply for.
+  /// \return Void.
+  void RecordResourceAcquirement(const NodeID &node_id,
+                                 const ResourceSet &required_resources);
+
+  /// Return acquired resources.
+  ///
+  /// \return Void.
+  void ReturnAcquiredResources();
+
+  GcsResourceManager &gcs_resource_manager_;
+
+ private:
+  absl::flat_hash_map<NodeID, std::list<ResourceSet>> acquired_resources_;
```

Comment:
`GcsScheduleStrategy` should better be stateless and have no side effect. 
we can move the GcsResourceManager operations out.  This can also avoid duplicated code.